design-proposals/klip-28-create-or-replace

```diff
@@ -0,0 +1,156 @@
+| | Scaling | Scaling upgrades change the physical properties of the query in order to enable better performance characteristics. |
+| | Unsupported | Unsupported upgrades are ones that will semantically change the query in an unsupported way. There are no plans to implement these migrations. |
+| Environment | Backfill | Backfill requires the output data to be accurate not just from a point in time, but from the earliest point of retained history |
+| | Cascading | Cascading environments contain queries that are not terminal, but rather feed into downstream stream processing tasks |
+| | Exactly Once | Exactly Once environments do not allow for data duplication or missed events |
+| | Ordered | Ordered environments require that a single offset delineates pre- and post-migration (no events are interleaved) |
+| | Live | Live environments describe queries that cannot afford downtime, either by means of acting as live storage (e.g. responding to pull queries) or feeding into high availability systems (powering important functionality) |
+
+### What is in scope
+
+- Specify a syntax that can support arbitrary upgrades
+- Design a validator to fail unsupported upgrades
+- Design a mechanism for upgrading queries under limited scope
+
+### What is not in scope
+
+The Design section below enumerates which upgrades are out of scope
+
+## Value/Return
+
+This KLIP will represent a significant step forward in the operability of ksqlDB in production, as
+noted in the background and motivation section.
+
+## Public APIS
+
+The syntax `CREATE OR REPLACE (STREAM | TABLE) source_name WITH (key=value, ...) AS query;` will be
+introduced to allow users to specify an existing stream or table to replace with a new query that
+will resume from the same processing point as any previously existing query.
+
+## Design
+
+If the `source_name` does not yet exist, a `CREATE OR REPLACE` statement functions identically to
+a normal `CREATE` statement. Otherwise, ksqlDB executes the following:
+
+1. Identify the original `queryID` that populates the source (`INSERT INTO` discussed later)
+2. Ensure the upgrade is valid
+3. Terminate `queryID`
+4. Start the new query under the same `queryID`
+
+A few changes need to happen in order to make this work. For 1, we need to maintain a mapping from
+source to queryID(s). If the source has multiple associated ids (in the case of `INSERT INTO`) then
+the upgrade will fail and not terminate any queries.
+
+For step 2, there will be a component to determine whether two topologies are "upgrade
+compatible"; the first iterations, which will be delivered as part of this KLIP, will only allow
+for the most basic upgrades: 
+
+- Any _transparent_ upgrade will be supported
+- Any _data selection_ upgrade will be supported
+- _Schema evolution_ upgrades will be supported on simple and stateful queries, but it will be
```

Comment:
Does this include _removing_ columns from the output schema, as well as _adding_ columns?

What about changing the _type_ of a column?

Removing and changing the type of existing columns has implications for downstream queries. The simplest solution would be to not allow either of these in this first pass, but not sure of the intent of the KLIP from this statement.  Can we make sure the doc is clearer on this please?

IMHO it's crucial that we ensure dumping the full database schema and running it on a fresh instance would succeed and result in all the same output schemas and data.

Reply:
For example, 

Let's say we have:

```sql
CREATE STREAM S1 AS 
   SELECT ID, NAME, COST FROM BLAH;

CREATE STREAM S2 AS
   SELECT ID, NAME FROM S1;
```

But then we run:

```sql
-- drops the `NAME` column from the projection:
CREATE OR REPLACE S1 AS
   SELECT ID, COST FROM BLAH;
```

Now S2 references a column in S1 that doesn't exist in the DB schema!  While this would work with formats that can support evolution and optional fields etc... it leaves the DB schema inconsistent, i.e. we lose referential integrity of the db schema.

Reply:
Personally, I'd go with not allowing rename / removing initially.  

Later we can build up system tables that track what columns are actually used in downstream queries and only allows you to remove / rename columns that aren't in use.

Reply:
Would be interesting to see what a trad db does if you try and rename/remove a column in DB if its used in a materialized view?

Reply:
@big-andy-coates I feel like this should follow the schema compatibility rules defined in schema registry, shouldn't it?
If schema registry isn't configured, maybe fall back to FORWARD, then one can only remove nullable fields, but I don't have a strong opinion on that

Reply:
@PeterLinder originally I was thinking along the same lines as you, but I think Andy has me convinced. While removing fields would be acceptable from a data schema evolution perspective, the table schema evolution would be broken by removing fields that are used in downstream queries.

Reply:
Added:

```
Note that _Schema Evolution_ compatibility is defined by the limitations of the serialization
format that is used with the added restrictions against removing fields and changing types to ensure
referential integrity of ksqlDB tables. This way, downstream query output schemas willn not be affected
by upstream schema evolution.
```

Reply:
Great! Though note its not just removing columns, but also renaming columns!

Reply:
From my perspective, there isn't be a difference between removing and renaming (renaming=removing and adding). If I'm checking that all columns from the previous schema exist in the new one, if one was renamed that check would fail!